### PR TITLE
Add support for values params for functions

### PR DESCRIPTION
**Pull request by dluc**
### Motivation and Context

1. When writing a prompt one might want to use the symbols `{{` and `}}` without triggering the template engine.
2. When invoking functions, one might want to pass a value rather than a variable, e.g. `{{ recall "my preferred browser" }}`.
3. Semantic functions might want to call other semantic functions.

### Description

This PR adds support for "values", so special syntax can be embedded, and so that functions can be invoked with predefined values without the need to create a temporary variable.

E.g.

```
{{ "{{" }} and {{ "}}" }} are special SK sequences.
```
```
... {{ read 'c:\documents\ai\file.txt' }} ...
```
```
... {{ "quotes' \"escaping\" example" }} ...
```
```
... {{ 'quotes\' "escaping" example' }} ...
```
```
{{ 'two special chars \\\' here' }}
```

Added missing tests for all the template language, including end to end integration tests covering several edge cases.

Also: semantic functions can now call other semantic functions. The previous implementation would allow semantic functions to call only native functions.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/src/SemanticKernel/SkillDefinition/IReadOnlySkillCollection.cs
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel/SkillDefinition/IReadOnlySkillCollection.cs:56]: Could be native, right? Same below.

File: dotnet/src/SemanticKernel/TemplateEngine/PromptTemplateEngine.cs
Comment (lemillermicrosoft) [inline dotnet/src/SemanticKernel/TemplateEngine/PromptTemplateEngine.cs:61]: May be out of scope of this review, but I'm seeing first hand and getting reports from folks that SK behaves differently than the playground with identical prompts and it seems to come down to '\r\n' vs '\n' - I'm wondering if we should go ahead and make that part of the engine to convert everything for consistency?

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/VarBlock.cs
Comment (shawncal) [inline dotnet/src/SemanticKernel/TemplateEngine/Blocks/VarBlock.cs:81]: nit string.Empty

File: dotnet/src/SemanticKernel/TemplateEngine/IPromptTemplateEngine.cs
Comment (shawncal) [inline dotnet/src/SemanticKernel/TemplateEngine/IPromptTemplateEngine.cs:37]: Can you fix this while you're in the code pls? ("Given a a...")

File: docs/PROMPT_TEMPLATE_LANGUAGE.md
Comment (dluc) [reply docs/PROMPT_TEMPLATE_LANGUAGE.md:206]: it's on purpose, see below line 208..210

File: dotnet/src/IntegrationTest/IntegrationTests.csproj
Comment (dluc) [reply dotnet/src/IntegrationTest/IntegrationTests.csproj:44]: probly not, checking

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/CodeBlock.cs
Comment (dluc) [reply dotnet/src/SemanticKernel/TemplateEngine/Blocks/CodeBlock.cs:13]: yeah it's on purpose, aside from test classes, I try to reduce the scope of these exception to a minimum

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/FunctionIdBlock.cs
Comment (dluc) [reply dotnet/src/SemanticKernel/TemplateEngine/Blocks/FunctionIdBlock.cs:27]: fixing - the code was throwing in IsValid, better doing it here

File: dotnet/src/IntegrationTest/IntegrationTests.csproj
Comment (dluc) [reply dotnet/src/IntegrationTest/IntegrationTests.csproj:40]: removed

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs
Comment (dluc) [reply dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs:50]: nice idea, thanks

Comment (dluc) [reply dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs]: done

**Review by lemillermicrosoft — APPROVED** [review]
Did not have chance to test yet, but LGTM.

File: dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs
Comment (shawncal) [inline dotnet/src/SemanticKernel/TemplateEngine/Blocks/Block.cs:39]: Got some new warnings here

**Review by shawncal — APPROVED** [review]

**Comment by shawncal** [comment]
@dluc Merge conflict (renamed tests folder) and a new warning. Happy to merge when ready

**Review by lemillermicrosoft — COMMENTED** [review]
Had one question about variable expansion on function name support.

File: docs/PROMPT_TEMPLATE_LANGUAGE.md
Comment (dluc) [reply docs/PROMPT_TEMPLATE_LANGUAGE.md:65]: correct, it's not supported. Variables can be used only to pass values, not to generate names/code. For more advanced coding, we rely on native functions

**Comment by dluc** [comment]
@shawncal @lemillermicrosoft warnings fixed / comments addressed

**Review by shawncal — APPROVED** [review]

**Pull request merged**
